File: JPP/Lista5/HASKEL/prime_factors.hs
import System.Environment (getArgs)

primeFactors :: Int -> [Int]
primeFactors n = factor n 2
  where
    factor n p
      | p * p > n      = [n | n > 1]
      | n `mod` p == 0 = p : factor (n `div` p) p
      | otherwise      = factor n (p + 1)

main :: IO ()
main = do
    args <- getArgs
    let n = read (head args) :: Int
        factors = primeFactors n
    putStrLn $ "Prime factors of " ++ show n ++ ":"
    print factors

File: JPP/Lista5/PROLOG/prime_factors.pl
is_prime(2).
is_prime(3).
is_prime(N) :-
    N > 3,
    N mod 2 =\= 0,
    \+ has_factor(N, 3).

has_factor(N, F) :-
    N mod F =:= 0.
has_factor(N, F) :-
    F * F < N,
    F2 is F + 2,
    has_factor(N, F2).

smallest_prime_factor(N, F) :-
    smallest_prime_factor(N, F, 2).

smallest_prime_factor(N, N, _) :- is_prime(N).
smallest_prime_factor(N, F, D) :-
    D * D =< N,
    (N mod D =:= 0 -> F = D ; D1 is D + 1, smallest_prime_factor(N, F, D1)).


prime_factors(1, []).
prime_factors(N, [F|Factors]) :-
    N > 1,
    smallest_prime_factor(N, F),
    N1 is N // F,
    prime_factors(N1, Factors).